core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java

```diff
@@ -15,74 +15,113 @@
 import alluxio.exception.ExceptionMessage;
 import alluxio.exception.FileDoesNotExistException;
 import alluxio.exception.InvalidPathException;
+import alluxio.master.file.meta.InodeTree.LockMode;
 import alluxio.master.file.meta.InodeTree.LockPattern;
 import alluxio.util.io.PathUtils;
 
 import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
 
 import java.io.Closeable;
+import java.util.ArrayList;
 import java.util.List;
 
 import javax.annotation.Nullable;
-import javax.annotation.concurrent.ThreadSafe;
+import javax.annotation.concurrent.NotThreadSafe;
 
 /**
- * This class represents a path of locked {@link Inode}, starting from the root.
+ * This class represents a locked path within the inode tree, starting from the root.
+ *
+ * The mUri and mPathComponents fields are immutable and contain the "target" path represented by
+ * the LockedInodePath.
+ *
+ * mLockList manages the locks held by the LockedInodePath.
+ *
+ * The mExistingInodes list holds inodes that are locked and known to exist. The inodes might not be
+ * individually locked - if one inode is write-locked, the inodes after it are implicitly
+ * write-locked and may be added to mExistingInodes. The inodes in mExistingInodes are a prefix of
+ * the inodes referenced by mPathComponents.
+ *
+ * If the full path exists, mExistingInodes will have size equal to mPathComponents.length after
+ * traversal. mExistingInodes is always at least as long as the list of inodes directly locked by
+ * mLockList.
+ *
+ * To create new inode paths from an existing path, use one of the lock*() methods. They return new
+ * locked inode paths that can be modified and closed without affecting the original path. Modifying
+ * the original inode path invalidates any subpaths that it has created.
+ *
+ * Locked inode paths are not threadsafe and should not be shared across threads.
  */
-@ThreadSafe
-public abstract class LockedInodePath implements Closeable {
+@NotThreadSafe
+public class LockedInodePath implements Closeable {
+  /** The root inode of the inode tree. */
```

Comment:
So will this always be the first inode in `mExistingInodes`? What is the "contract" between this `mRoot` variable and other lists in this class?

Reply:
After calling traverse(), this will always be the first inode in `mExistingInodes`. Updated the field documentation